Posts/2021/11/tfl-formatenumeration.md

```diff
@@ -0,0 +1,150 @@
+I set the limit to 99 in my profile file and that is usually more than sufficient.
+
+## Scoping of $FormatEnumerationLimit
+
+One interesting thing I found is that `$FormatEnumerationLimit` is scoped differently to my expectations.
+If you use a format command within a function/script (i.e. a child of the global scope), the command only uses the value from the global scope.
+
+The following code contains a function to illustrate the issue:
+
+```powershell
+function Test-FormatLimit1
+{
+  # Change format enum limit
+  "In Function, limit is: [$FormatEnumerationLimit]"
+  $FormatEnumerationLimit = 1
+  "After changing: [$FormatEnumerationLimit]"
+  Get-Process | Select-Object -Property Name, Threads -First 4
+}
+```
+
+You might think that this code would display the first thread in each of the first four processes. 
+You might, but you would be wrong, as you can see here:
+
+```powershell-console
+PS> # Here show the value and call the functin
+PS> "Before calling: [$FormatEnumerationLimit]"
+Before calling: [4]
+PS> Test-FormatLimit1
+In Function, limit is: [4]
+After changing: [1]
+
+Name                    Threads
+----                    -------
+AggregatorHost          {5240}
+ApplicationFrameHost    {16968, 2848}
+AppVShNotify            {9164}
+Atom.SDK.WindowsService {4064, 4908, 4912, 19144…}
+```
+
+As you can see from this output, the final process shows FOUR threads not ONE.
+This is because PowerShell seems to only use the globally scoped value, not the locally scoped copy.
+To get around this curious scoping, you can re-write the function like this:
+
+```powershell
+
+function Test-FormatLimit2
+{
+  # Change format enum limit Globally
+  $Old = $Global:FormatEnumerationLimit
+  $Global:FormatEnumerationLimit = 2
+  "After changing: [$Global:FormatEnumerationLimit]"
+  Get-Process | Select-Object -Property Name, Threads -First 4
+  # Change it back
+  $Global:FormatEnumerationLimit = $Old
+}
+```
+
+When you call the updated function, it now operates more as you might wish, like this:
+
+```powershell-console
+PS> # View the value
+PS> "Before calling: [$FormatEnumerationLimit]"
+Before calling: [4]#
+PS> # Now call the updated function
+PS> Test-FormatLimit2
+After changing: [2]
+
+Name                    Threads
+----                    -------
+AggregatorHost          {5240}
+ApplicationFrameHost    {16968, 2848}
+AppVShNotify            {9164}
+Atom.SDK.WindowsService {4064, 4908…}
+```
+
+So with some careful updating of the global variable, you can get the desired result.
+I am unclear whether this is a bug or a feature!
+In general, I teach my students to avoid manipulating global variables from within a script or a function (unless you know what you are doing).
+If you need to make changes to any global variable to make a function or script do what you want, ensure you know how to revert the variable to its original value.
```

Comment:
```suggestion
So, with some careful updating of the global variable, you can get the desired result.
In general, I teach my students to avoid manipulating global variables from within a script or a function (unless you know what you are doing).
If you need to make changes to any global variable to make a function or script do what you want, ensure you know how to revert the variable to its original value.

I am unclear whether this is a bug or a feature! To that end, I submitted a [feature request](https://github.com/PowerShell/PowerShell/issues/16360) in the PowerShell source repository. Feel free to add your opinion in the comments or upvote it if you want to see it added.
```